pkg/datarepair/checker/checker.go

```diff
@@ -217,3 +170,114 @@ func contains(a []string, x string) bool {
 	}
 	return false
 }
+
+func (checker *Checker) checkSegmentStatus(ctx context.Context, pointer *pb.Pointer, path string, monStats *durabilityStats) (err error) {
+	remote := pointer.GetRemote()
+	if remote == nil {
+		return nil
+	}
+
+	pieces := remote.GetRemotePieces()
+	if pieces == nil {
+		checker.logger.Debug("no pieces on remote segment")
+		return nil
+	}
+
+	missingPieces, err := checker.overlay.GetMissingPieces(ctx, pieces)
+	if err != nil {
+		checker.logger.Error("error getting missing pieces: ", zap.Error(err))
+		return err
+	}
+
+	monStats.remoteSegmentsChecked++
+	pathElements := storj.SplitPath(path)
+	if len(pathElements) >= 2 && pathElements[1] == "l" {
+		monStats.remoteFilesChecked++
+	}
+
+	numHealthy := int32(len(pieces) - len(missingPieces))
+	redundancy := pointer.Remote.Redundancy
+	// we repair when the number of healthy files is less than or equal to the repair threshold
+	// except for the case when the repair and success thresholds are the same (a case usually seen during testing)
+	if numHealthy >= redundancy.MinReq && numHealthy <= redundancy.RepairThreshold && redundancy.RepairThreshold != redundancy.SuccessThreshold {
+		if len(missingPieces) == 0 {
+			checker.logger.Warn("Missing pieces is zero in checker, but this should be impossible -- bad redundancy scheme.")
+			return nil
+		}
+		monStats.remoteSegmentsNeedingRepair++
+		err = checker.repairQueue.Insert(ctx, &pb.InjuredSegment{
+			Path:       path,
+			LostPieces: missingPieces,
+		})
+		if err != nil {
+			return err
+		}
+
+		// delete always returns nil when something was deleted and also when element didn't exists
+		_ = checker.irrdb.Delete(ctx, []byte(path))
+	} else if numHealthy < redundancy.MinReq {
+		// check to make sure there are at least *4* path elements. the first three
+		// are project, segment, and bucket name, but we want to make sure we're talking
+		// about an actual object, and that there's an object name specified
+		if len(pathElements) >= 4 {
+			project, bucketName, segmentpath := pathElements[0], pathElements[2], pathElements[3]
+			lostSegInfo := storj.JoinPaths(project, bucketName, segmentpath)
+			if contains(monStats.remoteSegmentInfo, lostSegInfo) == false {
+				monStats.remoteSegmentInfo = append(monStats.remoteSegmentInfo, lostSegInfo)
+			}
+		}
+
+		monStats.remoteSegmentsLost++
+		// make an entry in to the irreparable table
+		segmentInfo := &pb.IrreparableSegment{
+			Path:               []byte(path),
+			SegmentDetail:      pointer,
+			LostPieces:         int32(len(missingPieces)),
+			LastRepairAttempt:  time.Now().Unix(),
+			RepairAttemptCount: int64(1),
+		}
+
+		// add the entry if new or update attempt count if already exists
+		err := checker.irrdb.IncrementRepairAttempts(ctx, segmentInfo)
+		if err != nil {
+			checker.logger.Error("error handling irreparable segment to queue: ", zap.Error(err))
+			return err
```

Comment:
create new error, don't duplicate log entries by logging and returning error

Reply:
updated the code to return error with Error.New()